### DISCUSSION TITLE
main.py closes on its own

### DISCUSSION THREAD

**Discussion by tboy1337** [masterking32/MasterHamsterKombatBot#152] [Q&A] [answered]
Twice i have logged into the computer where the bot is running and found that main.py has closed itself, im using the newest version and config.

**Comment by tboy1337** [comment]
Ive tried using
```
from subprocess import run
from time import sleep

# Path and name to the script you are trying to start
file_path = "main.py" 

restart_timer = 2
def start_script():
    try:
        # Make sure 'python' command is available
        run("python "+file_path, check=True) 
    except:
        # Script crashed, lets restart it!
        handle_crash()

def handle_crash():
    sleep(restart_timer)  # Restarts the script after 2 seconds
    start_script()

start_script()
```
in another .py file to reopen main.py but it gets stuck at getting account data.

**Comment by tboy1337** [comment]
it closed again, definitely not windows updates or anything like that.

**Comment by tboy1337** [comment]
Ive also tried in AutoIT to reopen main.py
```
While True  
  If (ProcessExists("py.exe") = 0) Then  
    Run("python C:\PATH-TO-MAIN.PY\main.py")  
  EndIf  
  Sleep(1000)  
Wend
```
but that continually opens multiple instances of main.py for some reason.

**Comment by tboy1337** [comment]
its closed itself about 5 times today so far.

**Comment by ItsAmirAT** [comment]
same issue using ubuntu server

**Comment by tboy1337** [comment]
I'm glad it's not just me xD

**Comment by tboy1337** [comment]
the current code isnt creating a log file, ive gotten chatgpt to create some code which is working (its in https://github.com/masterking32/MasterHamsterKombatBot/discussions/147), from the logfile its getting to
```
[Master HamsterKombat Bot][INFO] 2024-08-12 17:01:48 [Account 1] Event registered successfully.
[Master HamsterKombat Bot][ERROR] 2024-08-12 17:01:49 [Account 1] Unable to get Bike Ride 3D in Hamster FAM key.
```
Thats when it closes.

**Comment by tboy1337** [comment]
you can use this code to reopen it for now untill there is a fix, create a new .py file in the same directory called anything you want and add this code:
```
import subprocess
import time

def restart_script(script_path):
    while True:
        print(f"Starting {script_path}...")
        process = subprocess.Popen(['python', script_path])
        process.wait()  # Wait for the script to exit
        print(f"{script_path} exited. Restarting in 5 seconds...")
        time.sleep(5)  # Optional delay before restarting

if __name__ == "__main__":
    script_to_run = 'main.py'  # Replace with the path to your script
    restart_script(script_to_run)
```
then run the new .py file.

**Comment by AmirhN1382** [comment, marked as answer]
Consider using Process Managers like ["PM2"](http://pm2.keymetrics.io/).
- yes it's for node.js, but works with python files too.
![image](https://github.com/user-attachments/assets/a9dd6a2d-2d29-4ba8-8067-00b24a7e8eaa)

**Comment by tboy1337** [comment]
@masterking32 this is a genuine issue, im not really sure why you have changed it to a discussion, the main.py is closing on its own when it hits a certain point.

**Reply by masterking32** [reply]
Sometimes you can use Google,
https://stackoverflow.com/questions/32127834/how-to-run-a-python-script-like-pm2-for-nodejs
`PM2` is fine you can use other programs for that, and that's not an issue. this is QA.
